packages/service-library/src/servicelib/archiving_utils/_interface_7zip.py

```diff
@@ -0,0 +1,330 @@
+_ALL_DONE_RE: Final[str] = r"Everything is Ok"
+
+_7ZIP_PATH: Final[Path] = Path("/usr/bin/7z")
+
+_TABLE_HEADER_START: Final[str] = "------------------- "
+_FILE_PERMISSIONS: Final[str] = " ..... "
+
+
+class ArchiveInfoParser:
+    def __init__(self) -> None:
+        self.total_bytes: NonNegativeInt | None = None
+        self.file_count: NonNegativeInt | None = None
+
+    async def parse_chunk(self, chunk: str) -> None:
+        # search for ` NUMBER bytes ` -> set byte size
+        if self.total_bytes is None and (match := re.search(_TOTAL_BYTES_RE, chunk)):
+            self.total_bytes = int(match.group(1))
+
+        # search for ` NUMBER files` -> set file count
+        if self.file_count is None and (match := re.search(_FILE_COUNT_RE, chunk)):
+            self.file_count = int(match.group(1))
+
+    def get_parsed_values(self) -> tuple[NonNegativeInt, NonNegativeInt]:
+        if self.total_bytes is None:
+            msg = f"Unexpected value for {self.total_bytes=}. Should not be None"
+            raise ArchiveError(msg)
+
+        if self.file_count is None:
+            msg = f"Unexpected value for {self.file_count=}. Should not be None"
+            raise ArchiveError(msg)
+
+        return (self.total_bytes, self.file_count)
+
+
+class ProgressParser:
+    def __init__(
+        self, decompressed_bytes: Callable[[NonNegativeInt], Awaitable[None]]
+    ) -> None:
+        self.decompressed_bytes = decompressed_bytes
+        self.total_bytes: NonNegativeInt | None = None
+
+        # in range 0% -> 100%
+        self.percent: NonNegativeInt | None = None
+        self.finished: bool = False
+        self.finished_emitted: bool = False
+
+        self.emitted_total: NonNegativeInt = 0
+
+    def _prase_progress(self, chunk: str) -> None:
+        # search for " NUMBER bytes" -> set byte size
+        if self.total_bytes is None and (match := re.search(_TOTAL_BYTES_RE, chunk)):
+            self.total_bytes = int(match.group(1))
+
+        # search for ` dd% ` -> update progress (as last entry inside the string)
+        if matches := re.findall(_PROGRESS_PERCENT_RE, chunk):
+            self.percent = int(matches[-1].strip().strip("%"))
+
+        # search for `Everything is Ok` -> set 100% and finish
+        if re.search(_ALL_DONE_RE, chunk, re.IGNORECASE):
+            self.finished = True
+
+    async def parse_chunk(self, chunk: str) -> None:
+        self._prase_progress(chunk)
+
+        if self.total_bytes is not None and self.percent is not None:
+            # total bytes decompressed
+            current_bytes_progress = int(self.percent * self.total_bytes / 100)
+
+            # only emit an update if something changed since before
+            bytes_diff = current_bytes_progress - self.emitted_total
+            if self.emitted_total == 0 or bytes_diff > 0:
+                await self.decompressed_bytes(bytes_diff)
+
+            self.emitted_total = current_bytes_progress
```

Comment:
question: why you do not directly use the ProgressBarData class? it contains all these things already.